accepted/0000-yomi-initial-integration.md

```diff
@@ -0,0 +1,361 @@
+
+The user can later decide to add the Yomi state into the `top.sls` for
+this system, fill the pillar data using the SaltForm provided in the
+package, and launch the installer by applying the highstate.
+
+## SaltForms and sub-forms
+
+Uyuni provides a mechanism, SaltForms, that allows the creation of the
+pillar data that suit the formula. The current relation between a
+SaltForm and a Salt state is 1:1, but using some empty `init.sls` we
+have the option of splitting a big SaltForm into multiple smaller
+SaltForms (subforms), that optionally belongs to the same group.
+
+The current `yomi-formula` is using this mechanism to split the pillar
+data into six sections:
+
+- yomi: general configuration data, like use `kexec` instead of
+  `reboot` after the installation, or some `systemd-firstboot` data.
+
+- yomi-bootloader: indicate the device where the bootloader will be
+  installed, and the parameters send to the kernel.
+
+- yomi-services: list of services that will be enabled, and is
+  `salt-minion` data will be copied into the new system.
+
+- yomi-software: repository registering and packages installation.
+
+- yomi-storage: device partition, LVM or RAID configuration, and file
+  systems applied.
+
+- yomi-users: creation of initial users, passwords and certificates.
+
+As we will see in future sections, those forms contains some
+overlapping with other salt-forms (the locale configuration that
+`systemd-firstboot` requires), and with other subsystems in Uyuni
+(like the software section).
+
+## Highstate and Workflow
+
+Installing a new system will be equivalent of applying a highstate to
+a system booted from the Yomi image.
+
+The current Yomi image is a JeOS based image with a `salt-minion`,
+some systemd overwrites that will help during the initial
+configuration, and the CLI tools required when a new installation is
+performed. You can find more information here:
+https://github.com/openSUSE/yomi#the-iso-image
+
+The current Yomi states are aware of differentiating when the system
+is booting from an external image (like the Yomi image), from when the
+boot comes from the local device. This is used to decide when the
+installation of the node makes sense or not when a highstate is
+re-applied.
+
+This means that we can set all the pillar data for the node, without
+risk of breaking or damaging already installed nodes when the Yomi
+state is in place for those systems.
+
+This open the question about the workflow that will be in place when a
+new system is allocated. The current prototype was tested like this
+one:
+
+1) Boot the new node via the Yomi ISO image. The local DHCP server
+   assign a FQDN to the system, and this is used automatically as a
+   minion ID.
+
+   (note: the master address was indicated via a Grub command line
+   that an overwrite in the image can read and interpret, how the new
+   minion will automatically discover the master is an open question
+   for me yet)
```

Comment:
In Retail it is done that the saltboot initrd minion asks for `salt` CNAME resolution. Then this FQDN is used as master. This place a requirement on customers DNS to have a proper CNAME for `salt` configured. See [saltboot.sh#L111](https://github.com/uyuni-project/retail/blob/master/kiwi-desc-saltboot/saltboot/suse-SLES12/root/saltboot.sh#L111)

Reply:
Yes, as is described the last option is to use `master` as a valid address form the master, I can move it to `salt` instead, but I wonder if there is a better option, like adding some metadata during the DHCP resolution, that a service can read and parse, and configure the minion accordingly.

Reply:
Yeah I know, that is what I replaced it (to avoid trivial hijacking of minions). I agree to revert to `salt` instead of `master` in the systemd overwrite that is executed before `salt-minion.service`, but I still think that there is a better solution that avoid default FQDN.

Reply:
To make things secure, we should ideally inject the master's public key into the RAM image and also enable strict checking. The image can also have the master's FQDN hardcoded. For non-PXE non-DNS/DHCP landscapes that's the only choice anyway.
Once we have to have managed DNS/DHCP anyway, I think just resolving the "salt" CNAME like saltboot is doing it isn't a bad approach.
We could of course also use SLP (Manager should still have that feature, which we never used) to find the Manager server on the network. But this doesn't sound more elegant than DNS.

Reply:
> To make things secure, we should ideally inject the master's public key into the RAM image and also enable strict checking.

I like the idea. I added it in the section "How the minion can find the master automatically"

> The image can also have the master's FQDN hardcoded. For non-PXE non-DNS/DHCP landscapes that's the only choice anyway.

In this case you can add the parameter 'master=master-address' in Grub, as is already in place.

> Once we have to have managed DNS/DHCP anyway, I think just resolving the "salt" CNAME like saltboot is doing it isn't a bad approach.

Yes, the image now is using 'salt' instead of master.